language/resource-api/README.md

```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
+      default: 'present',
+    },
+    name:        {
+      type:      'String',
+      desc:      'The name of the resource you want to manage.',
+      behaviour: :namevar,
+    },
+    some_array:  {
+      type: 'Optional[Array[String]]',
+      desc: 'An array aiding array attestation',
+    },
+    force: {
+      type:      'Boolean',
+      desc:      'If true, the specified array is inclusive and must match the existing state exactly.',
+      behaviour: :parameter,
+      default:   false,
+    },
+  },
+)
+
+# lib/puppet/provider/test_custom_insync/test_custom_insync.rb
+class Puppet::Provider::TestCustomInsync::TestCustomInsync < Puppet::ResourceApi::SimpleProvider
+  def insync?(context, name, property_name, is_hash, should_hash)
+    context.notice("Checking whether #{property_name} is out of sync")
+    case property_name
+    when :some_array
+      if should_hash[:force]
+        context.notice("Checking an order independent array")
+        return is_hash[property_name].sort == should_hash[property_name].sort
+      else
+        context.notice("Checking a subset match array")
+        found_members = (is_hash[property_name] & should_hash[property_name]).sort
+        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
+        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
+      end
+    end
+end
+```
+
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
+
+The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
+The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.
+
+The `insync?` method's return value will drive the behavior of Puppet for that property:
+
+- If `insync?` returns `nil`, property comparison is handed back to Puppet's built-in property insync checking.
+  - This means that _only_ properties which actually need custom insync logic need to be handled in the `insync?` method on the provider. Everything else can be ignored, which will return `nil` and use default comparison logic.
+- If `insync?` returns `true`, the property is treated as insync.
+- If `insync?` returns `false`, the property is treated as out of sync and will surface the default Puppet change reporting.
+  - e.g. `changed <is value> to <should value>`
+- If `insync?` returns a string, the property is treated as out of sync and the string is surfaced in the Puppet change reporting.
```

Comment:
```suggestion
- If `insync?` returns a string, the property is treated as out of sync and the string is used for change reporting instead of the default. This can be used to provide additional information about the reason for a change.
```